src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -635,73 +637,78 @@ private CancellationTokenRegistration RegisterCancellation(CancellationToken can
         // TODO: Remove this overload once https://github.com/dotnet/roslyn/issues/17287 is addressed
         // and the compiler doesn't lift the span temporary from the call site into the async state
         // machine in debug builds.
-        private void ParseStatusLine(ArraySegment<byte> line, HttpResponseMessage response) =>
+        private static void ParseStatusLine(ArraySegment<byte> line, HttpResponseMessage response) =>
             ParseStatusLine((Span<byte>)line, response);
 
-        private void ParseStatusLine(Span<byte> line, HttpResponseMessage response)
+        private static void ParseStatusLine(Span<byte> line, HttpResponseMessage response)
         {
             // We sent the request version as either 1.0 or 1.1.
             // We expect a response version of the form 1.X, where X is a single digit as per RFC.
 
-            const int MinStatusLineLength = 12;     // "HTTP/1.1 123" 
-
-            if (line.Length < MinStatusLineLength ||
-                line[0] != 'H' ||
-                line[1] != 'T' ||
-                line[2] != 'T' ||
-                line[3] != 'P' ||
-                line[4] != '/' ||
-                line[5] != '1' ||
-                line[6] != '.' ||
-                line[8] != ' ')
+            // Validate the beginning of the status line and set the response version.
+            const int MinStatusLineLength = 12; // "HTTP/1.x 123" 
+            if (line.Length < MinStatusLineLength || line[8] != ' ')
             {
                 ThrowInvalidHttpResponse();
             }
 
-            // Set the response HttpVersion
-            byte minorVersion = line[7];
-            response.Version =
-                minorVersion == '1' ? HttpVersion.Version11 :
-                minorVersion == '0' ? HttpVersion.Version10 :
-                !IsDigit(minorVersion) ? throw new HttpRequestException(SR.net_http_invalid_response) :
-                new Version(1, minorVersion - '0');
+            ulong first8Bytes = BitConverter.ToUInt64(line);
+            if (first8Bytes == s_http11Bytes)
+            {
+                response.SetVersionUnchecked(HttpVersion.Version11);
+            }
+            else if (first8Bytes == s_http10Bytes)
+            {
+                response.SetVersionUnchecked(HttpVersion.Version10);
+            }
+            else
+            {
+                byte minorVersion = line[7];
+                line[7] = (byte)'1'; // Enable checking the first 7 bytes easily by comparing against HTTP/1.1
```

Comment:
It's a little weird to modify the buffer.  This code isn't perf critical, would it be better to just do something dumb here?

Reply:
Yeah, went back and forth on that.  It shouldn't matter that we're modifying it, but I could do the "simpler" thing and just add the per-char comparisons here... unfortunately that simpler thing is actually longer, but whatever :smile:

Reply:
Yeah... to make it smaller, we could possibly just have a static byte[] for "HTTP/1." and do some sort of sequence based comparison (is there a Span.SequenceEquals?)

Reply:
There is.  I'll do that.